[API8] Add datapack support

mattmess1221 opened the pull request:
I'm assuming the client's resource manager was reused for the server. Likely more edits will come when I have access to deobfuscated code.

mattmess1221 committed:
Add resource manager for data packs

Addd a ResourcePath for easier path validation
Move package to resources instead of pack

Vanilla datapacks can be disabled

Move package to resource (remove plural)
Add builders
Complete* javadocs
Add functionality for dynamic resources

More work on resources.
Resource creation is more foolproof
Added ResourceId annotation for resource injection
Various other additions and fixes

Use a packid instead of a name.

Lift more methods into ResourceProvider and PackProvider
Add some QoL methods into ResourcePath (lifted from java.nio.Path)

Add a walkResources method

Add ResourceReloadEvent.Post

Add a resourcetype to be used later

liach commented:
should we merge resource path with catalog key?

mattmess1221 commented:
If we do, we should settle on a better suited name. It doesn't make sense to get a `Resource` from a `CatalogKey`, and neither does a `CatalogType` with `ResourcePath`

Though they are slightly different. While both have a namespace and value, the resource path can contain a forward slash `/` and dot `.`. If anything, it should have a common parent-interface indicating it can have a namespace.

liach commented:
Something like `BinomialName` for a parent interface?

mattmess1221 commented:
I was thinking it would include the word namespace. Something like `NamespacedIdentifier`

liach commented:
`SpacedName`

kashike commented on src/main/java/org/spongepowered/api/Game.java:
```diff
@@ -35,6 +35,7 @@
 import org.spongepowered.api.event.CauseStackManager;
 import org.spongepowered.api.event.EventManager;
 import org.spongepowered.api.network.ChannelRegistrar;
+import org.spongepowered.api.resources.ResourceManager;
```
Singular package names - `resource`.

ryantheleach commented:
I don't like that this outright deprecates the Asset API.

The goals of the Asset API were to simplify the process for new programmers to extract resources from their jar files, to their plugin directories on disk, usually used for extracting default or example configurations.

If this is to happen, I'd like at least a little attention paid to giving those people a clear migration path, and maybe some docs to go along with it to explain exactly what this is and how it works.
  

I also think the name DataPack is pretty close to the Mojang defined term.

If this isn't intended to be directly compatible with that, or interact with that in some way then it's poorly named.

But maybe I'm just not following why the name change is needed?

ryantheleach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,245 @@
+     *
+     * @return The parent pack.
+     */
+    Pack getPack();
```
Do Resources always have Packs?

liach replied:
I think there may be a default pack for minecraft

mattmess1221 replied:
If used properly, there are no orphans in the resource manager. The only way for this to return null is for the developer to use `Resource.builder()` and not add it to a pack.

I'm actually working on something that makes this a bit more foolproof. Standby.

liach commented on src/main/java/org/spongepowered/api/Sponge.java:
```diff
@@ -152,10 +154,15 @@ public static EventManager getEventManager() {
         return check(assetManager);
     }
 
+    public static ResourceManager getResourceManager() {
```
jdoc

liach commented on src/main/java/org/spongepowered/api/resource/InputStreamSupplier.java:
```diff
@@ -0,0 +1,45 @@
+     * create a new input stream.
+     *
+     * @return A new input stream
+     * @throws IOException if an error occurs
```
Is there a case regulation for `@throws` tags?

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,245 @@
+     * @return The list of strings
+     * @throws IOException if an error occurs
+     */
+    default List<String> readLines(Charset charset) throws IOException {
```
Can we get a java stream here? We can use collector to collect it and do more complex operations

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,132 @@
+         * @return This builder
+         * @throws IllegalArgumentException if the namespace contains illegal characters
+         */
+        Builder namespace(String namespace) throws IllegalArgumentException;
```
Can we use a plugin object to supply the namespace, too?

mattmess1221 replied:
Yes, I just added a method for that.

liach commented on src/main/java/org/spongepowered/api/resource/Pack.java:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return The name
+     */
+    Text getName();
```
Do we only have text names? String names are easier to identify compared to text ones (remember that we get a list of active packs from the resource provider and plugins may want to disable some of them)

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/Pack.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @return The name
+     */
+    String getName();
```
This is a `String` but the `Builder` uses a `Text` object?
Should this name also be used in the `registerPack` method? If so, the name argument in the method is unnecessary.

Cybermaxke commented:
Please keep the `ResourcePath` separate from the `CatalogId`. They represent two different things and shouldn't be merged as one. They also have different contraints.

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceManager.java:
```diff
@@ -0,0 +1,119 @@
+     * @param name
+     * @param pack
+     */
+    void registerPack(String name, Pack pack);
```
I think that it's fine to keep this method in here.

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceManager.java:
```diff
@@ -0,0 +1,119 @@
+     * @return The pack which was unregistered or empty if it wasn't registered
+     */
+    Optional<Pack> unregisterPack(String name);
+
```
Maybe also `prioritizeAfter`, `prioritizeBefore`, `prioritizeFirst` and `prioritizeLast` methods?

liach replied:
Can't we just arrange that list?

mattmess1221 replied:
The way I was thinking is the pack order would be determined by calling `reload(List<Pack>)`.  Though I also have an idea about using Comparable to sort the packs.

Cybermaxke replied:
What happens if someone passes in a `Pack` that isn't registered?

Cybermaxke replied:
What happens if you try to unregister a plugin pack?

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -88,6 +88,9 @@ static ResourcePath parse(String path) throws IllegalArgumentException {
     String getPath();
 
+    @Override
+    int hashCode();
```
Why is this in the API?

liach replied:
For arbitrary order?

mattmess1221 replied:
I put it there originally because I wanted it to be clear that this can be used in a Map. Now I'm just saying so in the docs.

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceProvider.java:
```diff
@@ -38,4 +39,18 @@
+     * @return A stream of resource paths
+     */
+    Stream<ResourcePath> walk(ResourcePath path);
+
```
I wouldn't go with the `walk` name, it can be confusing when using in combination with `ResourceManager`. You can also return the `Resource` directly when walking or just add variants for `Resource`s and `ResourcePath`s. A depth could also be useful to limit the directory structure complexity:
`Stream<ResourcePath> listResourcePaths(ResourcePath resourcePath);`
`Stream<ResourcePath> walkResourcePaths(ResourcePath resourcePath, int maxDepth);`
`Stream<ResourcePath> walkResourcePaths(ResourcePath resourcePath);`
`Stream<Resource> walkResources(ResourcePath resourcePath, int maxDepth);`
`Stream<Resource> walkResources(ResourcePath resourcePath);`

mattmess1221 replied:
What would be the difference between `listResourcePaths` and `walkResourcePaths`?

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/PackProvider.java:
```diff
@@ -1,16 +1,75 @@
+import java.util.Optional;
 
-@FunctionalInterface
 public interface PackProvider {
```
Any reason why these methods are moved outside of the `ResourceManager`?

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -92,6 +93,54 @@ static ResourcePath parse(String path) throws IllegalArgumentException {
+     *
+     * @return The root path
+     */
+    ResourcePath getRoot();
```
Root would return `namespace:`, path is empty, if the docs are true this would return the parent path. Add a separate `getParent` method for this.

liach replied:
This method does not appear meaningful to me. Why?

liach commented on src/main/java/org/spongepowered/api/resource/ResourceId.java:
```diff
@@ -24,13 +24,24 @@
  *
- * The type this should be used on is {@code Optional<Resource>}.
+ * The type this should be used on is {@code Optional<Resource>}. When applied
+ * to a method, the {@link Resource} will be
```
will be what?

Cybermaxke commented on src/main/java/org/spongepowered/api/event/resource/ResourceReloadEvent.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    interface Pre extends ResourceEvent {
+
+    }
```
No `Post` event?

mattmess1221 replied:
I don't think having `onReload(ResourceReloadEvent.Post)` is good api. If someone were to do `ResourceReloadEvent` instead, it would receive both events, which is probably not what they meant.

Cybermaxke replied:
The `Pre` event is called before the resources are reloaded, how do you it's finished reloading?

stephan-gh commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,157 @@
+     * @throws IOException if an error occurs
+     */
+    default Reader getReader(Charset charset) throws IOException {
+        return new InputStreamReader(openStream(), charset);
```
I'd return a `BufferedReader` here. You'll likely almost always want a buffered reader, and if you don't you'll know to create it from `openStream()` manually.

mattmess1221 committed:
Resource.getReader now returns a BufferedReader

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,202 @@
+     *
+     * @return The list of parents
+     */
+    List<ResourcePath> getParents();
```
Also a regular `getParent`?

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceTypes.java:
```diff
@@ -0,0 +1,15 @@
+import org.spongepowered.api.util.generator.dummy.DummyObjectProvider;
+
+// TODO I don't like this
+public class ResourceTypes {
```
I wouldn't add this in the API, this seems more like a implementation detail to me. The only reason to still use the `assets` package is because it's used by the Asset API, the assets package in vanilla will only be used for textures, etc. Maybe just use the `assets` package for the Asset API and the `data` package for the resource api?

Cybermaxke replied:
Since we are also going to support the client, `ResourceType` can stay

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceType.java:
```diff
@@ -0,0 +1,8 @@
+package org.spongepowered.api.resource;
```
Missing header

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceTypes.java:
```diff
@@ -0,0 +1,15 @@
+package org.spongepowered.api.resource;
```
Missing header

mattmess1221 committed:
ResourcePath is iterable on it's parents

mattmess1221 committed:
Fixup some javadoc

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -94,36 +95,38 @@ static ResourcePath parse(String path) throws IllegalArgumentException {
+     * @return The list of paths
      */
-    List<String> getNames();
+    List<ResourcePath> getParts();
```
IMO the `getNames` and `getParents` methods were good, but `getParts` doesn't seem right. A resource path cannot be split into resource path parts, because it also contains the namespace. If it would split into stings that would make sense, but then it would just be the `getNames` method.

Javadocs still says it's the parents?

Cybermaxke replied:
I already noticed that you fixed the javadoc.

mattmess1221 committed:
Add some filename related methods

mattmess1221 committed:
Add a method in Pack.Builder for providing from a file

mattmess1221 committed:
Add some methods to ResourceReloadEvent.Pre for manipulating the packs to load

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,254 @@
+     * @see #getBaseName()
+     * @see #getExtension()
+     */
+    String getFileName();
```
I wouldn't call this `getFileName` since there is no guarantee that it's actually a file, it can also be a directory. This also breaks the `getBaseName` method.

mattmess1221 replied:
I disagree. Yes, there is no guarantee that it is a file, but that's not up to the path to decide. The path has no say in whether something exists, or is a file (or directory).

And `getBaseName` was named after `FilenameUtils.getBaseName`, which I added because commons-io is not included in the API.

Lignium commented on src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java:
```diff
@@ -0,0 +1,39 @@
+import org.spongepowered.api.event.Event;
+import org.spongepowered.api.resource.ResourceManager;
+
+public interface ResourceEvent extends Event {
```
Javadoc?

liach commented:
Can we move datapacks to each world, given that is how they are handled on both client and servers? Current structure won't work for clients.

mattmess1221 commented:
@liach Nope. Resources are not per-world. There are only 2 instances of `SimpleReloadableResourceManager` in 1.13. They are located in `Minecraft.resourceManager` and `MinecraftServer.resourceManager`

mattmess1221 committed:
Rework the resource system to be more inline with vanilla.
I'm still working on figuring out how ResourcePackList works.

Cybermaxke commented:
Resources are basically per save file, on the server there is only one. On the client its the one when you load a world (including all dimensions).

liach commented:
@killjoy1221 However, data packs are not loaded as resources; they are never included in the resource manager. They are simply associated with worlds.

mattmess1221 commented:
I haven't looked too closely on how datapacks work yet, but I'm pretty sure the world will be added as a data pack, which can also be removed. I'm going to have to study the code to figure how how everything works with `PackFinder` and `ResourcePackList`

mattmess1221 committed:
Add a PackRepository and PackInfo. The PackInfo is used to get the Pack,
but the PackInfo cannot be gotten back from the pack.
Temp remove ResourceId until I can figure out a good way to use it.
Remove the try-with-resources from Resource. It should be done from the caller.

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceManager.java:
```diff
@@ -45,7 +49,7 @@
      * @return The resource
      */
-    Optional<Resource> getResource(ResourcePath path);
+    Resource getResource(ResourcePath path) throws IOException;
```
Maybe rename this to `openResource` and add a method `hasResource` to check whether a resource exists with a given path.

Cybermaxke commented on src/main/java/org/spongepowered/api/resource/ResourceManager.java:
```diff
@@ -0,0 +1,79 @@
+     * @return A collection of resource paths
+     */
+    Collection<ResourcePath> getResources(String path, Predicate<String> filter);
+
```
Maybe it's still worth adding some of the methods from my lantern datapacks pr?
https://github.com/LanternPowered/LanternServer/pull/53/files#diff-d8831f2ec678210c3618a5b72bbd5a47R47

mattmess1221 replied:
Changing this into a `Stream` may be useful.

mattmess1221 committed:
Update some javadocs and modify a few methods

liach commented on src/main/java/org/spongepowered/api/event/resource/ResourceReloadEvent.java:
```diff
@@ -0,0 +1,63 @@
+
+    /**
+     * Called before the {@link ReloadableResourceManager} is reloaded. At this
+     * point, the {@link PackRepository#getEnabledPacks() enables packs} can be
```
`enabled packs`

liach commented on src/main/java/org/spongepowered/api/item/inventory/ItemStack.java:
```diff
@@ -85,7 +85,7 @@ static ItemStack of(ItemType itemType, int quantity) {
      */
     static ItemStack empty() {
-        return builder().itemType(ItemTypes.NONE).build();
+        return builder().itemType(ItemTypes.AIR).build();
```
Should put this in a separate pr to prevent merge conflicts

Cybermaxke replied:
The PR should be rebased on top of the latest changes.

liach commented on src/main/java/org/spongepowered/api/plugin/PluginContainer.java:
```diff
@@ -131,15 +133,26 @@ default String getName() {
+     * @return The plugin's resource pack.
+     */
+    default PackInfo getPack() {
+        return Sponge.getServer().getPackRepository().getPack(this);
```
As far as I know, resource packs and data packs have different information, even as they can share the same pack mcmeta file. This prevents a plugin from having both packs, which is the case for forge mods, etc.

mattmess1221 replied:
This will probably have to wait for a full impl until the client api gets fleshed out a bit more. As it is now, `ResourceTypes.ASSETS` is restricted to the client, and `DATA` to the server.

liach replied:
I suggest removing this method. Plugin Containers may exist without the server. This method itself is default, meaning that the removal will not reduce functionality.

liach replied:
Instead of removing this method, we can leave it abstract so that implementations are correct.

liach replied:
We should consider moving this to the client and server interfaces instead.

mattmess1221 committed:
Add resourcetype argument to PluginContainer.getPack()
Also some javadocs

liach commented:
@killjoy1221 no, data packs can be in integrated servers

liach commented on src/main/java/org/spongepowered/api/resource/PackRepository.java:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @return The enabled packs
+     */
+    Collection<PackInfo> getEnabledPacks();
```
Mention that the iteration order or the collection is the packs' priority,

liach commented on src/main/java/org/spongepowered/api/resource/PackRepository.java:
```diff
@@ -0,0 +1,85 @@
+     * @param packs The active packs
+     * @see ReloadableResourceManager#reload()
+     */
+    void setActivePacks(Collection<PackInfo> packs);
```
Mention that the iteration order of the `packs` is the priority of active packs.

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+ *     try (Resource res = resourceManager.getResource(path)) {
+ *         logger.info(res.readString(DefaultCharsets.UTF_8));
+ *     } catch (IOException e) {
+ *         e.printStackTrace();
```
We should recommend users to log the error or throw it instead of `printStackTrace`

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,245 @@
+ * be located roughly at {@code data/namespace/path}</p>
+ *
+ * <p>A resource path should be usable in a {@link Map}, so implementations
+ * should override {@link #hashCode()} and {@link #equals(Object)}.</p>
```
Put this implnote in the method javadocs as well.

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,245 @@
+ * <p>A resource path should be usable in a {@link Map}, so implementations
+ * should override {@link #hashCode()} and {@link #equals(Object)}.</p>
+ */
+public interface ResourcePath extends Comparable<ResourcePath>, Iterable<ResourcePath> {
```
The iterable, in my opinion, is very rarely used and can cause mod conflicts like SpongePowered/SpongeForge#2540

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,245 @@
+ * contain any special characters or uppercase letters.
+ *
+ * <p>In the pack, the path will point to a resource. The resource should
+ * be located roughly at {@code data/namespace/path}</p>
```
Do you need to note a transformation of resource path between user and actual usage? Users supply advancement id `minecraft:story/root` that will be translated to `minecraft:advancements/story/root.json`, which locates `data/minecraft/advancements/story/root.json`.

mattmess1221 replied:
This is specific to resources. A better class for that one would be `NamespacedId`. There's a conversation about this.

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,245 @@
+     * @param path The path to use
+     * @return A new ResourcePath
+     */
+    static ResourcePath of(String namespace, String path) throws IllegalArgumentException {
```
How about other `minecraft(String)` or `plugin(PluginContainer/Object, String)` methods?

mattmess1221 replied:
I don't feel those methods are necessary. You can just as easily stick in `"minecraft`" or `"foobar"`. I prefer it to be explicit for this.

liach commented:
@killjoy1221 Can you target 1.14 instead?

liach commented on src/main/java/org/spongepowered/api/resource/ResourceTypes.java:
```diff
@@ -0,0 +1,38 @@
+
+    public static final ResourceType ASSETS = DummyObjectProvider.createFor(ResourceType.class, "ASSETS");
+
+    public static final ResourceType DATA = DummyObjectProvider.createFor(ResourceType.class, "DATA");
```
It's not supposed to be like this... this is pack type. resource types are like `advancements` `loot_tables` `functions` `tags/functions` `tags/fluids` `models` `textures` `blockstates` etc...

mattmess1221 committed:
Change for 1.14
Adds async reloaders
Tweak some method names
Remove some less useful methods
TODO: Javadocs

liach commented on src/main/java/org/spongepowered/api/resource/AsyncReloader.java:
```diff
@@ -0,0 +1,34 @@
+        return Sponge.getRegistry().createBuilder(Builder.class);
+    }
+
+    ResourceManager getResourceManager();
```
I don't think async reloaders are manager aware, are they?

mattmess1221 replied:
They are.

liach commented on src/main/java/org/spongepowered/api/resource/ResourceManager.java:
```diff
@@ -48,17 +52,26 @@
+     * @throws FileNotFoundException If no resources exist
      */
-    List<Resource> getResources(ResourcePath path);
+    List<Resource> getResources(ResourcePath path) throws IOException, FileNotFoundException;
```
Remember to document the order of the list is based on priority. Higher priority entries are placed earlier in this list (I think this is the case; I may be wrong).

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackInfo.java:
```diff
@@ -86,4 +80,22 @@
+
+    Compatibility getCompatability();
+
+    enum Priority {
```
`InsertionPosition`

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackInfo.java:
```diff
@@ -86,4 +80,22 @@
+    enum Compatibility {
+        OLD,
+        NEW,
+        OK
```
Consider using some other thing like a string version, like what we do for `GameVersion`

mattmess1221 committed:
Fix the build

Cybermaxke commented on src/main/java/org/spongepowered/api/Engine.java:
```diff
@@ -25,12 +25,16 @@
  * Shared functionality between {@link Client} and {@link Server} engines.
  */
-public interface Engine {
+public interface Engine extends Executor {
```
You can't make engine a executor, all tasks need to be scheduled through the scheduler to know which plugin submitted them. You can however get a exectutor from the scheduler for a specific plugin.

mattmess1221 replied:
Minecraft and MinecraftServer are already executors. Adding a separate method would essentially be returning this. But I guess that's an implementation detail.

Cybermaxke replied:
We need to track which plugin posts a task, vanilla mc does not. So we can't just expose this in the API.

Cybermaxke commented:
@killjoy1221 Can you rebase on top of `api-8`?

mattmess1221 committed:
Fix some rebase artifacts

mattmess1221 committed:
AsyncReloader had too much impl.

liach commented on src/main/java/org/spongepowered/api/resource/ResourcePath.java:
```diff
@@ -0,0 +1,246 @@
+         * @param plugin The owning plugin
+         * @return This builder
+         */
+        Builder namespace(PluginContainer plugin);
```
Accept object as well for plugin instance?

mattmess1221 replied:
I was keeping consistent with `CatalogKey.Builder`

Zidane replied:
We don't do Object plugin parameters anymore.

mattmess1221 committed:
Pack should be closeable.

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackRepository.java:
```diff
@@ -0,0 +1,85 @@
+     *
+     * @return The enabled packs
+     */
+    Collection<PackInfo> getEnabledPacks();
```
Again, make it a list to show that enabled packs are ordered

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackInfo.java:
```diff
@@ -0,0 +1,83 @@
+/**
+ * Basic info about a {@link Pack}
+ */
+public interface PackInfo extends DataSerializable, TextRepresentable {
```
Why `DataSerializable`?

liach replied:
Also, return the pack type as well.

mattmess1221 replied:
It has no concept of `PackType`. I really don't remember why I made it data serializable.

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackInfo.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     * @return The display name
+     */
+    Text getDisplayName();
```
More exactly, this is just a name pack repo gives and with the desc in the hover

mattmess1221 replied:
Should I use `String getId()` and `Text getName()`?

liach replied:
I suggest you add a `getName` method additionally that returns the name used to refer to this pack info for commands.

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackRepository.java:
```diff
@@ -0,0 +1,85 @@
+     * @param packs The active packs
+     * @see Engine#reloadResources()
+     */
+    void setEnabledPacks(Collection<PackInfo> packs);
```
`Iterable` or `List` is better; the enabled packs are ordered, which plain iterable and list can convey.

liach commented on src/main/java/org/spongepowered/api/resource/pack/Pack.java:
```diff
@@ -72,14 +72,21 @@
      * Gets the metadata of this pack. The {@link DataView} represented is of
      * the pack.json file in the pack root. If the pack does not contain a
-     * pack.json, {@link Optional#empty()} is returned.
+     * pack.json, it is an error.
```
wtf, it is `pack.mcmeta`, and if that one is absent, a default one will be returned

mattmess1221 replied:
And what should a default mcmeta contain?

liach replied:
It just contains an assumed pack version and a broken message in its description

liach commented on src/main/java/org/spongepowered/api/event/resource/ResourceEvent.java:
```diff
@@ -22,14 +22,34 @@
+     * Fired after a {@link ResourceManager} has been reloaded. This can be
+     * used to reload some systems that are resource based.
+     */
+    interface Reload extends ResourceEvent {
```
Please allow plugins to work async! This will increase load on the server and cannot be profiled by vanilla loading bar, etc.

mattmess1221 committed:
Fixes to javadocs and nitpicks

liach commented on src/main/java/org/spongepowered/api/resource/pack/PackTypes.java:
```diff
@@ -0,0 +1,38 @@
+
+    // SORTFIELDS:ON
+
+    public static final PackType ASSETS = DummyObjectProvider.createFor(PackType.class, "ASSETS");
```
Nah, this is actually resource.
![image](https://user-images.githubusercontent.com/7806504/57576981-da7eda00-7421-11e9-8cf3-8dd96e59c0b5.png)

mattmess1221 replied:
Yeah, but I don't really like that name.

liach replied:
`RESOURCE` is already an established name. (I don't ask you to name this one with the client prefix) This name is not mapped, but rather inferred from the enum constructor parameters passed.

Well, minecraft wiki documents resource and data packs, not assets and data packs. Beware of that.

mattmess1221 committed:
Change PackTypes.ASSETS to RESOURCES

mattmess1221 committed:
Remove the assets package.

mattmess1221 committed:
Fix some nitpicks

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     * @param charset The charset
+     * @return A new reader
+     */
+    default BufferedReader getReader(Charset charset) {
```
`createReader`/`makeReader`/`newReader` as `get` implies this is an intrinsic property.

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     * @throws IOException if an error occurs
+     */
+    default byte[] readBytes() throws IOException {
+        return ByteStreams.toByteArray(getInputStream());
```
Resource leak here.

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return A new input stream
+     */
+    InputStream getInputStream();
```
`createInputStream`/`newInputStream`/`makeInputStream`

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     *
+     * @return The path
+     */
+    ResourcePath getPath();
```
In my opinion, `getName` sounds better. See how dinnerbone refers to `<namespace>:<name>` in https://minecraft.net/en-us/article/minecraft-snapshot-17w43a

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     * @throws IOException if an error occurs
+     */
+    default String readString(Charset charset) throws IOException {
+        return CharStreams.toString(getReader(charset));
```
Resource leak

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Reads a list of string lines from this resource using the given charset.
```
Note that the stream needs to be closed.

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     */
+    default DataView readDataView(DataFormat format) throws IOException {
+        try {
+            return format.readFrom(getInputStream());
```
Possible resource leak; unfortunately sponge api did not document if the stream will be closed after usage.

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     * @throws IOException if an error occurs
+     */
+    default void copyTo(Path path, OpenOption... options) throws IOException {
+        MoreFiles.asByteSink(path, options).writeFrom(getInputStream());
```
Possible leak?

liach commented on src/main/java/org/spongepowered/api/resource/Resource.java:
```diff
@@ -0,0 +1,178 @@
+     * @throws IOException if an error occurs
+     */
+    default void copyTo(@WillNotClose OutputStream out) throws IOException {
+        ByteStreams.copy(getInputStream(), out);
```
Need to close the input stream and note that the output stream is not flushed.

liach commented on src/main/java/org/spongepowered/api/resource/pack/Pack.java:
```diff
@@ -0,0 +1,99 @@
+ * A pack can contain several {@link Resource Resources}. Each pack is
+ * independently loaded with a configured priority.
+ */
+public interface Pack extends Closeable {
```
Mention that a pack can be used both as a data and resource pack, like that pack from the vanilla client.

liach reviewed (changes requested):
Must fix all those resource leaks before this merge.

mattmess1221 committed:
Remove WillNotClose from Resource copy.

mattmess1221 committed:
Change getInputStream javadoc to represent current code.

mattmess1221 commented:
InputStream leakage is covered by `Resource#close()` and should be handled by the caller.

Zidane commented:
@killjoy1221 Need this finished up soon else it won't make it for API 8.0.0

mattmess1221 commented:
#2135
